csrc/scheduler/normalization_inner_outer.cpp

```diff
@@ -199,31 +440,108 @@ std::pair<int64_t, int64_t> getPersistentBufferSize(
-      ? persistent_buffer_size_info.persistent_buffer_size
-      : std::min(
-            persistent_buffer_size_info.persistent_buffer_size,
-            persistent_buffer_size_info.projected_persistent_buffer_size);
+  // Project to inputs when there is at least one outer broadcast tensor or
+  // projected persistent buffer size is smaller. When projecting to inputs, the
+  // outer broadcast tensor is reused in the loop over the iteration dimension,
+  // test shows it is faster than the non-projected version which requires
+  // reload from gmem for each iteration.
+  // Note: in current use cases (layer norm bwd and RMS norm bwd), there are
+  // outer broadcast tvs and always project to inputs.
+  const auto& outer_broadcast_tvs = getOuterBroadcastTvs(fusion, reduction_tvs);
+  buffer_params.project_to_input =
+      normalization_scheduler_utils::isProjectBufferToInputs(
+          fusion,
+          runtime_info,
+          persistent_buffer_info,
+          persistent_buffer_size_info,
+          ScheduleHeuristic::InnerOuterPersistent,
+          outer_broadcast_tvs.empty());
 
-  // in combined_inner_outer_reduction, the partial results of outer
-  // reductions must be persistent, allow register spill avoid segmentation
-  std::vector<TensorView*> outer_reduction_tvs;
-  for (auto tv : reduction_tvs) {
-    if (!scheduler_utils::isFastestDimReduction(tv)) {
-      outer_reduction_tvs.emplace_back(tv);
+  auto total_buffer_size = buffer_params.project_to_input
+      ? persistent_buffer_size_info.projected_persistent_buffer_size
+      : persistent_buffer_size_info.persistent_buffer_size;
+  total_buffer_size +=
+      partialOuterReductionBufferSize(reduction_tvs, runtime_info);
+
+  const auto dev_prop = at::cuda::getCurrentDeviceProperties();
+  int64_t smem_overhead = scheduler_utils::getSharedMemoryOverheadPerBlock(
+      fusion,
+      reduction_tvs,
+      InnerOuterPersistentKernelScheduler::threads_per_block_max);
+  int64_t available_smem =
+      (int64_t)dev_prop->sharedMemPerMultiprocessor - smem_overhead;
+  int64_t available_regs = scheduler_utils::register_file_size_full;
+  buffer_params.smem_overhead = smem_overhead;
+
+  // Put all the persistent tensors in registers
+  buffer_params.regs_buffer_size = total_buffer_size;
+  buffer_params.smem_buffer_size = 0;
+
+  // Relocate buffers to shared memory until the buffer size in registers is
+  // within the allowable limit.
+  if (buffer_params.regs_buffer_size > available_regs) {
+    const auto sorted_candidate_tvs = buffer_params.project_to_input
+        ? sortProjectableBufferInputs(
+              fusion,
+              persistent_buffer_info.persistent_buffers,
+              persistent_buffer_info.projectable_buffer_inputs,
+              outer_broadcast_tvs,
+              reduction_tvs,
+              runtime_info)
+        : sortPersistentBuffers(persistent_buffer_info.persistent_buffers);
+
+    // determine the number of buffers to move to shared memory
```

Comment:
Give more details how the sorted tensor list is going to be used to figure out this. Looks like the algorithm is to visit each tensor from the largest to the smallest and try to store as many largest tensors as possible in registers. Is that right?

Reply:
The visiting order is from small to large. I think maybe the `i-1` index is confusing and it is revised.
Added more comments and revised the code.

Two steps:
```
  // (3.1) Sort the candidate persistent buffers, based on its read/write usage.
  // Put the buffer that is less frequently used in the front.
```
and 
```
  // (3.2) Before this loop, all buffers are in registers.
  // Try to move buffer from register to shared memroy. The visiting order is
  // from the first buffer to the last buffer in the sorted_candidate_tvs.
  // After one buffer is moved to shared memory, the buffer size in registers
  // and shared memory are updated accordingly. Break if required register and
  // shared memory are lower than limit or shared memory exceeds the limit.
```

Reply:
Yeah, when I saw that and also `<= N` instead of usual `< N`, I thought the order was reversed. Sorry for my confusion.